libeos-payg/fake-clock.c

```diff
@@ -106,18 +106,82 @@ epg_fake_clock_set_time (EpgFakeClock *self,
   self->time_secs = new_time;
 }
 
+typedef struct {
+  GSource parent;
+
+  EpgFakeClock *clock; /* owned */
+  gint64 ready_time;
+} EpgFakeSource;
+
+static gboolean
+epg_fake_source_check (GSource *source)
+{
+  EpgFakeSource *self = (EpgFakeSource *) source;
+
+  return epg_clock_get_time (EPG_CLOCK (self->clock)) >= self->ready_time;
+}
+
+static gboolean
+epg_fake_source_dispatch (GSource    *source,
+                          GSourceFunc callback,
+                          gpointer    user_data)
+{
+  EpgFakeSource *self = (EpgFakeSource *) source;
+
+  if (callback == NULL)
+    {
+      g_warning ("Fake source dispatched without callback. "
+                 "You must call g_source_set_callback().");
+      return G_SOURCE_REMOVE;
+    }
+
+  self->ready_time = G_MAXINT64;
+
+  return callback (user_data);
+}
+
+static void
+epg_fake_source_finalize (GSource *source)
+{
+  EpgFakeSource *self = (EpgFakeSource *) source;
+
+  g_clear_object (&self->clock);
+}
+
+static const GSourceFuncs epg_fake_source_funcs = {
+  .check = epg_fake_source_check,
+  .dispatch = epg_fake_source_dispatch,
+  .finalize = epg_fake_source_finalize,
+};
+
+/**
+ * epg_fake_clock_source_new_seconds:
+ * @clock: an #EpgClock
+ * @interval: the timeout interval in seconds
+ * @error: a return location for #GError
+ *
+ * This creates a GSource which will be ready when @interval seconds have
+ * passed according to epg_fake_clock_get_time().
+ *
+ * Returns: (transfer full) (nullable): the newly created #GSource
+ */
 static GSource *
 epg_fake_clock_source_new_seconds (EpgClock  *clock,
                                    guint      interval,
                                    GError   **error)
 {
   EpgFakeClock *self = (EpgFakeClock *)clock;
-  guint interval_clamped;
+  g_autoptr(GSource) source = NULL;
+  EpgFakeSource *fake_source = NULL;
 
   g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), NULL);
 
-  interval_clamped = MIN (interval, G_MAXUINT / MSEC_PER_SEC);
-  return epg_boottime_source_new (interval_clamped * MSEC_PER_SEC, error);
+  source = g_source_new ((GSourceFuncs *)&epg_fake_source_funcs,
+                         sizeof (EpgFakeSource));
+  fake_source = (EpgFakeSource *)source;
+  fake_source->clock = g_object_ref (self);
+  fake_source->ready_time = epg_clock_get_time (EPG_CLOCK (self)) + interval;
```

Comment:
Nitpick: Would be good to put in an assertion to make sure this addition doesn’t overflow.